pyrate/core/dem_error.py

```diff
@@ -0,0 +1,307 @@
+    nrows, ncols = ifg.shape
+    lon = np.zeros((nrows, ncols))  # pre-allocate 2D numpy array
+    lat = np.zeros((nrows, ncols))  # pre-allocate 2D numpy array
+    for i in range(0, nrows): # rows are y-direction
+       for j in range(0, ncols): # cols are x-direction
+           lon_ij, lat_ij = convert_pixel_value_to_geographic_coordinate(j, i, transform)
+           lon[i, j] = lon_ij
+           lat[i, j] = lat_ij
+
+    return lon, lat
+
+
+def get_radar_coords(ifg, params):
+    """
+    Function to get radar coordinates for each pixel in the multi-looked interferogram dataset.
+    Radar coordinates are identical for each interferogram in the stack.
+    """
+    # lookup table file:
+    lookup_table = params[cf.LT_FILE]
+    # PyRate IFG multi-looking factors
+    ifglksx = params[cf.IFG_LKSX]
+    ifglksy = params[cf.IFG_LKSY]
+    print(lookup_table)
+    # transform float lookup table file to np array
+    lt_az, lt_rg = read_lookup_table(ifg, lookup_table, ifglksx, ifglksy)
+    # replace 0.0 with NaN
+    lt_az[lt_az==0.0] = np.nan
+    lt_rg[lt_rg == 0.0] = np.nan
+
+    return lt_az, lt_rg
+
+
+def calc_local_geometry(ifg, ifg_path, az, rg, lon, lat, params):
+    """
+    Function to calculate local look angle, incidence angle and geodetic azimuth for each pixel.
+    """
+    # read relevant metadata from first IFG
+    a = float(ifg.meta_data[ifc.PYRATE_SEMI_MAJOR_AXIS_METRES])
+    b = float(ifg.meta_data[ifc.PYRATE_SEMI_MINOR_AXIS_METRES])
+    se = float(ifg.meta_data[ifc.PYRATE_SAR_EARTH_METRES])
+    near_range = float(ifg.meta_data[ifc.PYRATE_NEAR_RANGE_METRES])
+    rps = float(ifg.meta_data[ifc.PYRATE_RANGE_PIX_METRES])
+    heading = float(ifg.meta_data[ifc.PYRATE_HEADING_DEGREES])
+
+    # convert angles to radians
+    lon = lon / 180 * np.pi
+    lat = lat / 180 * np.pi
+    heading = heading / 180 * np.pi
+
+    # Earth radius at given latitude
+    re = np.sqrt(np.divide(np.square(a**2 * np.cos(lat)) + np.square(b**2 * np.sin(lat)), \
+                           np.square(a * np.cos(lat)) + np.square(b * np.sin(lat))))
+
+    # range measurement at pixel ij -> this requires the mli.par as input since rps would be different in the slc.par
+    range = near_range + rps*rg
+
+    # look angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    look_angle = np.arccos(np.divide(se**2 + np.square(range) - np.square(re), 2 * se * range))
+
+    # incidence angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    incidence_angle = np.pi - np.arccos(np.divide(np.square(range) + np.square(re) - se**2, 2 * np.multiply(range, re)))
+    # for validation with GAMMA output only
+    incidence_angle_gamma = np.pi / 2 - incidence_angle
```

Comment:
move this out of prod code and in to a test for validating against Gamma output

Reply:
To be done once we've got a new test data set with the required GAMMA outputs. Commented out for now with a todo note.

Reply:
To be done once a test data set with all relevant input for the baseline/geometry functionality is implemented. For now I've commented the calculation and output of *_angle_gamma files out and made a todo comment on moving this into a test.